src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs

```diff
@@ -316,9 +336,92 @@ private string FormatEnumValueToString(string value, JavaScriptEncoder? encoder)
             return converted;
         }
 
+        private static string[] SplitFlagsEnum(string value)
+        {
+            // todo: optimize implementation here by leveraging https://github.com/dotnet/runtime/issues/934.
+            return value.Split(
+#if BUILDING_INBOX_LIBRARY
+                ValueSeparator
+#else
+                new string[] { ValueSeparator }, StringSplitOptions.None
+#endif
+                );
+        }
+
         internal override T ReadAsPropertyNameCore(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
         {
+            bool success = TryParseEnumQuick(ref reader, options, out T value, out string? _);
+            if (!success)
+            {
+                ThrowHelper.ThrowJsonException();
+            }
+
+            return value;
+        }
+
+        private T ReadEnumFromString(ref Utf8JsonReader reader, JsonSerializerOptions options)
+        {
+            bool success = TryParseEnumQuick(ref reader, options, out T value, out string? enumString);
+
+            if (success)
+            {
+                return value;
+            }
+
+            if (_namingPolicy == null)
+            {
+                ThrowHelper.ThrowJsonException();
+            }
+
+            if (enumString == null && reader.TokenType != JsonTokenType.Null)
+            {
+                enumString = reader.GetString();
+            }
+
+            if (enumString == null)
+            {
+                ThrowHelper.ThrowJsonException();
+            }
+
+            Debug.Assert(_enumValueCache != null, "Enum value cache should be instantiated if a naming policy is specified.");
+
+            if (!enumString.Contains(ValueSeparator))
+            {
+                success = _enumValueCache.TryGetValue(enumString, out value);
+            }
+            else
+            {
+                string[] enumValues = SplitFlagsEnum(enumString);
+                ulong result = 0;
+
+                for (int i = 0; i < enumValues.Length; i++)
+                {
+                    success = _enumValueCache.TryGetValue(enumValues[i], out value);
+                    if (!success)
+                    {
+                        break;
+                    }
+
+                    result += ConvertToUInt64(value);
```

Comment:
Shouldn't this be?
```suggestion
                    result |= value;
```
I think `result` could simply be `T` so that removes the need for any intermediate conversions.

Reply:
Yeah I tried this earlier, but it doesn't seem to work with the generic `<T>`. I'm getting [CS0019](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs0019?f1url=%3FappId%3Droslyn%26k%3Dk(CS0019)).

Reply:
I see. Even using `long` though shouldn't this be `|=` still? It's not the same as `+=` so this would result in weird outputs if for example I'm passing the same id twice:
```C#
public enum Foo { A = 1 }

ulong x;
x += (ulong)Foo.A;
x += (ulong)Foo.A;
(Foo)x; // 2

x |= (ulong)Foo.A;
x |= (ulong)Foo.A;
(Foo)x; // A
```

Reply:
Ah yes missed that.